File: 0053-MaximumSubarray/Solution.java
import java.util.*;

public class Solution {
    
    public int maxSubArray(int[] nums) {
        int len = nums.length;
        int prev = nums[len - 1]; // last element
        int max = prev;
        for (int i = len - 2; i >= 0; i--) { // starting at second to last element
            int curr = nums[i];
            if (prev > 0) {
                curr += prev;
            }
            prev = curr;
            max = Math.max(max, curr);
        }
        return max;
    }

    public static void main(String[] args) {
        System.out.println();
    }

}

File: 0053-MaximumSubarray/solution.py
class Solution:

    def maxSubArray(self, nums: List[int]) -> int:
        res = prev = nums[0]
        for n in islice(nums, 1, None):  # start=1, continue until the end
            curr = max(prev, 0) + n
            res = max(res, curr)
            prev = curr
        return res


if __name__ == "__main__":
    s = Solution()

